File: cpp/KalkCpp/Shape/Shape2D/Polygon/Quadrilateral/square.cpp
#include <iostream>
#include "square.h"
#include "math.h"
#include "typeinfo"

Square::Square(const std::vector<Point2D>& vp2d, ColorRGB* color): Quadrilateral(vp2d, "Square", color)
{
    if(!canBuild(vp2d))
    {
        points = buildSquare(1);
    }
}


Square* Square::clone() const
{
    return new Square(*this);
}

Square::Square(double sideLength, ColorRGB* color): Quadrilateral(buildSquare(sideLength), "Square", color)
{

}

bool Square::canBuild(double sideLenght) const
{
 return sideLenght > 0;
}

bool Square::canBuild(const std::vector<Point2D>& vp2d) const
{
    if(vp2d.size() == 4)
    {
        if(Point2D::distance(vp2d[1], vp2d[0]) > 0)
        {
            return true;
        }
    }
    return false;
}

std::vector<Point2D> Square::buildSquare(double sideLength) const
{
    if(!canBuild(sideLength))
        sideLength = 1;
    std::vector<Point2D> points =  std::vector<Point2D>();

    points.push_back(Point2D());
    points.push_back(Point2D(sideLength));
    points.push_back(Point2D(sideLength,sideLength));
    points.push_back(Point2D(0,sideLength));

    return points;
}



Square* Square::operator-(const Shape2D& s) const
{
    double newSide = 0;
    Square* newSquare = nullptr;
    if(typeid(s) == typeid(const Square&))
    {
        const Square& sq = static_cast<const Square&>(s);
        newSide = base() - sq.base();
        newSquare = new Square(buildSquare(newSide), getColor()->operator -(*s.getColor()));
    }else{
        newSide = base() - s.sideToOperator();
        newSquare = new Square(buildSquare(newSide), getColor()->operator -(*s.getColor()));
    }
    return newSquare;
}

Square* Square::operator+(const Shape2D& s) const
{
    double newSide = 0;
    Square* newSquare = nullptr;
    if(typeid(s) == typeid(const Square&))
    {
        const Square& sq = static_cast<const Square&>(s);
        newSide = base() + sq.base();
        newSquare = new Square(buildSquare(newSide), getColor()->operator +(*s.getColor()));
    }else{
        newSide = base() + s.sideToOperator();
        newSquare = new Square(buildSquare(newSide), getColor()->operator +(*s.getColor()));
    }
    return newSquare;
}

Square* Square::operator*(const Shape2D& s) const
{
    double newSide = 0;
    Square* newSquare = nullptr;
    if(typeid(s) == typeid(const Square&))
    {
        const Square& sq = static_cast<const Square&>(s);
        newSide = base() * sq.base();
        newSquare = new Square(buildSquare(newSide), getColor()->operator *(*s.getColor()));
    }else{
        newSide = base() * s.sideToOperator();
        newSquare = new Square(buildSquare(newSide), getColor()->operator *(*s.getColor()));
    }
    return newSquare;
}

Square* Square::operator/(const Shape2D& s) const
{
    double newSide = 0;
    Square* newSquare = nullptr;
    if(typeid(s) == typeid(const Square&))
    {
        const Square& sq = static_cast<const Square&>(s);
        newSide = base() / sq.base();
        newSquare = new Square(buildSquare(newSide), getColor()->operator /(*s.getColor()));
    }else{
        newSide = base() / s.sideToOperator();
        newSquare = new Square(buildSquare(newSide), getColor()->operator /(*s.getColor()));
    }
    return newSquare;
}

double Square::sideToOperator() const
{
    return base();
}



double Square::base() const
{
    return Point2D::distance(getPoints()[1], getPoints()[0]);
}



double Square::height() const
{
    return base();
}


double Square::diagonal() const
{
    return base() * sqrt(2);
}


double Square::area() const
{
    return pow(base(),2);
}

File: Java/src/Shape/Shape2D/Polygon/Quadrilateral/Square.java
package Shape.Shape2D.Polygon.Quadrilateral;

import Color.ColorRGB;
import Shape.Point2D;
import Shape.Shape2D.Shape2D;

import java.util.ArrayList;

public class Square extends Quadrilateral {

    public Square(ArrayList<Point2D> p) {
        super(p, "Square");
        if (!canBuild(p)) {
            points = buildSquare(1);
        }
    }

    public Square(ArrayList<Point2D> p, ColorRGB color) {
        super(p, "Square", color);
        if (!canBuild(p)) {
            points = buildSquare(1);
        }
    }

    public Square(double sideLength) {
        super(null, "Square");
        points = buildSquare(sideLength);
    }

    public Square(double sideLength, ColorRGB color) {
        super(null, "Square", color);
        points = buildSquare(sideLength);

    }

    public Square(Square s) {
        super(s);
    }

    private boolean canBuild(double sideLength) {
        return sideLength > 0;
    }

    private boolean canBuild(ArrayList<Point2D> vp2d) {
        if (vp2d.size() == 4) {
            return Point2D.distance(vp2d.get(1), vp2d.get(0)) > 0;
        }
        return false;
    }

    private ArrayList<Point2D> buildSquare(double sideLength) {
        if (!canBuild(sideLength))
            sideLength = 1;

        ArrayList<Point2D> points = new ArrayList<>();

        points.add(new Point2D());
        points.add(new Point2D(sideLength));
        points.add(new Point2D(sideLength, sideLength));
        points.add(new Point2D(0, sideLength));

        return points;
    }

    public double sideToOperator() {
        return base();
    }

    public double base() {
        return Point2D.distance(points.get(1), points.get(0));
    }

    public double height() {
        return base();
    }

    public double diagonal() {
        return base() * Math.sqrt(2);
    }

    public double area() {
        return Math.pow(base(), 2);
    }

    public Square sum(Shape2D other) {
        double newSide;
        Square newSquare;
        if (getClass() == other.getClass()) {
            Square s = (Square) other;
            newSide = base() + s.base();
            newSquare = new Square(buildSquare(newSide), getColor().sum(other.getColor()));
        } else {
            newSide = base() + other.sideToOperator();
            newSquare = new Square(buildSquare(newSide), getColor().sum(other.getColor()));
        }
        return newSquare;
    }

    public Square sub(Shape2D other) {
        double newSide;
        Square newSquare;
        if (getClass() == other.getClass()) {
            Square s = (Square) other;
            newSide = base() - s.base();
            newSquare = new Square(buildSquare(newSide), getColor().sub(other.getColor()));
        } else {
            newSide = base() - other.sideToOperator();
            newSquare = new Square(buildSquare(newSide), getColor().sub(other.getColor()));
        }
        return newSquare;
    }

    public Square mul(Shape2D other) {
        double newSide;
        Square newSquare;
        if (getClass() == other.getClass()) {
            Square s = (Square) other;
            newSide = base() * s.base();
            newSquare = new Square(buildSquare(newSide), getColor().mul(other.getColor()));
        } else {
            newSide = base() * other.sideToOperator();
            newSquare = new Square(buildSquare(newSide), getColor().mul(other.getColor()));
        }
        return newSquare;
    }

    public Square div(Shape2D other) {
        double newSide;
        Square newSquare;
        if (getClass() == other.getClass()) {
            Square s = (Square) other;
            newSide = base() / s.base();
            newSquare = new Square(buildSquare(newSide), getColor().div(other.getColor()));
        } else {
            newSide = base() / other.sideToOperator();
            newSquare = new Square(buildSquare(newSide), getColor().div(other.getColor()));
        }
        return newSquare;
    }


    public String toString() {
        return super.toString() + "Side: " + base() + "\nDiagonal: " + diagonal() +
                "\nArea: " + area() + "\nArea: " + area() + "\nPerimeter: " + perimeter();
    }

}
